feat!: make `Container::exec` synchronous and return output

DDtKey opened the pull request:
Closes #344

Remove of `-d` argument for `Container::exec` to ensure command has finished.
It's extremely useful when you need to ensure that command was executed as expected, but container logs can't tell this.

It also now returns output of command.

Any comments are welcome, that's quite straightforward way to implement it.

DDtKey commented:
@thomaseizinger I apologize for being annoying 😅 but I'd really appreciate for reviewing/merging

thomaseizinger reviewed (changes requested):
Thanks for the ping and sorry for the long delay, I am only passively maintaining this library.

I am not sure I fully understand your usecase. Can you give me an example of what you want to do?

The commands are currently executed synchronously, meaning you can be sure that they executed once `Container::exec` returned. Why would we need another list of `WaitFor`s (apart from the fact that this isn't designed for arbitrary commands anyway)?

thomaseizinger commented:
Are you trying to just solve https://github.com/testcontainers/testcontainers-rs/issues/344 or do you have your own usecase you want to solve here?

DDtKey commented:
@thomaseizinger yes, it solves #344

> The commands are currently executed synchronously, meaning you can be sure that they executed once Container::exec returned.

That's actually not true.  Currently (in the main branch) it uses `.arg("-d")`, so it doesn't wait. So I removed that flag and added additional checks of output if you would need it.

For example I want wait until there is a message `success` in stdout OR `failed` in stderr (too simplified example though), but it's helpful in order to make sure conditions are passed without manual checks of output (moreover, `exec` doesn't return output)

DDtKey commented:
Agree 👍

DDtKey commented on testcontainers/src/core/logs.rs:
```diff
@@ -36,21 +36,19 @@ impl<'d> LogStreamAsync<'d> {
 
-impl fmt::Debug for LogStream {
+impl<R: Read> fmt::Debug for LogStream<R> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         f.debug_struct("LogStream").finish()
     }
 }
 
-impl LogStream {
-    pub fn new(stream: impl Read + 'static) -> Self {
-        Self {
-            inner: Box::new(stream),
-        }
+impl<R: Read> LogStream<R> {
+    pub fn new(stream: R) -> Self {
+        Self { inner: stream }
```
It's a bit unrelated because I reverted changes of `WaitFor`, but decided to keep it because it removes extra `Box`ing without breaking any backward compatibility

thomaseizinger replied:
I am pretty sure this is a breaking change still. I'd like to keep the boxing to keep the API simpler and the type encapsulated.

DDtKey replied:
🤔  looks like it's `pub(crate)` and not exposed. 

I can revert ofc, but seems to be not breaking one

DDtKey commented on testcontainers/src/core/container.rs:
```diff
@@ -213,6 +225,13 @@ where
     }
 }
 
+/// Represents an output of `exec` command.
+/// `stdout` & `stderr` is represented as bytes because it might contain non-utf chars and responsibility should be on the caller end.
+pub struct ExecOutput {
+    pub stdout: Vec<u8>,
+    pub stderr: Vec<u8>,
+}
```
Created separate type, because tuple with the same types could be confusing and erroneous. 

Also decided not to convert it to string, because it's up to caller if output can contain non-utf8

DDtKey commented:
I can squash commits if needed after review

thomaseizinger commented on testcontainers/src/core/container.rs:
```diff
@@ -182,18 +189,23 @@ where
     }
 
-    pub fn exec(&self, cmd: ExecCommand) {
+    pub fn exec(&self, cmd: ExecCommand) -> ExecOutput {
```
It is unfortunately still a breaking change but that is okay. Can you add a changelog entry please?

DDtKey replied:
Yeah. you're right. Sure

thomaseizinger reviewed (commented):
Thanks for amending the PR. Some comments!

thomaseizinger commented on testcontainers/src/core/container.rs:
```diff
@@ -182,18 +183,23 @@ where
     }
 
-    pub fn exec(&self, cmd: ExecCommand) {
+    pub fn exec(&self, cmd: ExecCommand) -> ExecOutput {
```
Can you replicate this API for the `ContainerAsync` type please?

DDtKey replied:
`DockerAsync` doesn't have `exec`. Should I add this?

thomaseizinger replied:
> `ContainerAsync` & `DockerAsync` doesn't have `exec`. Should I add this? 

Oh okay. We really need some tests to keep these APIs on par.

All good then, can you open an issue? Thanks!

DDtKey replied:
Sure, created #475

thomaseizinger commented on testcontainers/src/clients/cli.rs:
```diff
@@ -418,7 +421,7 @@ impl Docker for Cli {
                 }
                 WaitFor::Duration { length } => {
-                    std::thread::sleep(length);
+                    sleep(length);
```
Let's keep the diff minimal please.

DDtKey replied:
AFAIR it was highlighted by clippy locally - because 'extra' export, but sure, not related to this PR

thomaseizinger reviewed (commented):
Thanks!
